target/linux/ramips/mt7621/base-files/sbin/ubnt_erx_migrate.sh

```diff
@@ -0,0 +1,86 @@
+#!/bin/sh
+
+. /lib/functions.sh
+
+include /lib/upgrade
+
+. /usr/share/libubox/jshn.sh
+
+BOARD="$(board_name | sed 's/,/_/g')"
+
+# Use snapshots until a stable 24.x release
+SITE="https://downloads.openwrt.org/snapshots/targets/ramips/mt7621"
+FILE="openwrt-ramips-mt7621-${BOARD}-squashfs-sysupgrade.bin"
+tar_file="/tmp/sysupgrade.img"
+SITE=${TESTSITE:-$SITE}
+
+
+confirm_migration() {
+    case "$(board_name)" in
+        ubnt,edgerouter-x|\
+        ubnt,edgerouter-x-sfp)
+            compat=$(uci -q get system.@system[0].compat_version)
+            if [ "$compat" != "1.1" ]; then
+                echo "Incompatible compat version $compat" >&2
+                exit 1
+            fi
+            ;;
+        *)
+            echo "Incompatible board $(board_name)" >&2
+            exit 1
+            ;;
+    esac
+
+    echo -e "\033[0;31mWARNING:\033[0m This script will migrate your OpenWrt system to a new layout as"
+    echo "required for the linux 6.6 kernel. This process will erase all your current settings."
+    echo "It is recommended to back up your system before proceeding."
+    echo ""
+
+    read -p "Do you want to proceed with the migration? (y/n)" confirm
+    if [ "$confirm" != "y" ]; then
+        echo "Migration canceled."
+        exit 0
+    fi
+}
+
+download_image(){
+    wget -qO "$tar_file" "$SITE/$FILE"
+    sha256=$(wget -qO- "$SITE/sha256sums" | grep "$BOARD-squashfs-sysupgrade.bin" | cut -d ' ' -f1)
+    sha256local=$(sha256sum "$tar_file" | cut -d ' ' -f1)
+    if [ "$sha256" != "$sha256local" ]; then
+            echo "Downloaded image checksum mismatch" >&2
+            exit 1
+    fi
+
+    board_dir=$( (tar tf "$tar_file" | grep -m 1 '^sysupgrade-.*/$') 2> /dev/null)
+    export board_dir="${board_dir%/}"
+
+    tar xC /tmp -f "$tar_file" 2> /dev/null
+    if [ ! -f /tmp/$board_dir/kernel ] || [ ! -f /tmp/$board_dir/root ]; then
+        echo "Invalid image file" >&2
+        exit 1
+    fi
+    cp /sbin/ubnt_erx_stage2.sh /tmp
+
+}
+
+confirm_migration
+download_image
+
```

Comment:
Would it work, if you did 
- `prepare_kernel` here,
- On the running system, modified `platform.sh` for edgerouter to something like `"CI_KERNPART="none"; nand_do_upgrade $1"`
- just call `sysupgrade` with the downloaded upgrade package

Just so that you utilise the existing sysupgrade functionality

Reply:
I was working under the assumption, based on ynezz comments above, I can't modify files in `/lib/upgrade`. I do agree it could be simplified by doing so however, by either patching `platform.sh` or `ubnt.sh` on the running system. I guess however the changes will persist even if sysupgrade fails for some reason (which it probably can't since need to use -F)

@ynezz Is something like msandber mentions acceptable to backport to stable release? or do the migration scripts need to be kept standalone as I have done so far?